Add support for depth function in spatial materials

apples opened the pull request:
Adds "Depth Function" property to spatial materials, which controls the depth comparison operator.

Resolves https://github.com/godotengine/godot-proposals/issues/1298.

Test project: https://github.com/apples/DepthFunctionTest

Seems to work fine in all three rendering modes in my simple testing, but more testing is likely necessary as this touches core rendering logic. I'm a little unsure about the GLES3 implementation, but I tried my best.

Currently only useful for alpha pass materials, due to https://github.com/godotengine/godot/issues/73158.

![image](https://user-images.githubusercontent.com/2352020/219801224-c7326415-8b2c-4dfd-8392-144cb36244ac.png)

Ansraer commented on servers/rendering/renderer_rd/forward_clustered/scene_shader_forward_clustered.cpp:
```diff
@@ -250,8 +261,36 @@ void SceneShaderForwardClustered::ShaderData::set_code(const String &p_code) {
+				break;
+		}
 	}
 	bool depth_pre_pass_enabled = bool(GLOBAL_GET("rendering/driver/depth_prepass/enable"));
```
Just want to point out that for opaque materials #70214 will overwrite the compare op you just set later on.
This is currently not a problem due to #73158 (which sends custom no-depth draw materials into the alpha pass) but could be a potential problem in the future.

apples replied:
As things are now, it would be difficult to integrate depth functions into the opaque render pass anyways, due to potential sorting issues. I'm currently looking into solutions for this, but it might be beyond the scope of this PR.

apples commented on servers/rendering/renderer_rd/forward_clustered/scene_shader_forward_clustered.cpp:
```diff
@@ -250,8 +261,36 @@ void SceneShaderForwardClustered::ShaderData::set_code(const String &p_code) {
 		depth_stencil_state.enable_depth_write = depth_draw != DEPTH_DRAW_DISABLED ? true : false;
+		switch (depth_function) {
+			case DEPTH_FUNCTION_LESS_OR_EQUAL:
+				depth_stencil_state.depth_compare_operator = RD::COMPARE_OP_LESS_OR_EQUAL;
```
I agree. Done. I think I used a switch to be able to have a `default` case, but that's unnecessary here.

QbieShay commented:
Hey, what's the state of this?

Calinou reviewed (approved):
Tested locally, it works as expected in all 3 rendering methods.

**Testing project:** [test_pr_73527.zip](https://github.com/godotengine/godot/files/12299794/test_pr_73527.zip)

It's possible to get a transparent x-ray effect to work with an opaque material if you use **Grow** with `-0.001` as a value in the next pass:

![image](https://github.com/godotengine/godot/assets/180032/3a3c7378-9aba-4ff3-a611-bd36f3cf5d08)

However, this is a fiddly setup that is difficult to figure out and Z-fighting will occur in the distance. A custom shader that sets grow distance based on the distance to the camera might work better. (Some outline effects may benefit from a fixed size regardless of camera distance, so it could be worth considering as a core material feature.)

If this caveat is well-documented in the StandardMaterial3D class reference and manual page, this should still be a net usability upgrade overall. However, I wonder if the depth function property should be hidden for opaque materials that have **Depth Draw Mode** set to a value other than **Never**, as it appears to do nothing otherwise.

Calinou commented on doc/classes/BaseMaterial3D.xml:
```diff
@@ -150,6 +150,9 @@
 			Determines when depth rendering takes place. See [enum DepthDrawMode]. See also [member transparency].
 		</member>
+		<member name="depth_function" type="int" setter="set_depth_function" getter="get_depth_function" enum="BaseMaterial3D.DepthFunction" default="0">
+			Determines which comparison operator is used when testing depth. See [enum DepthFunction].
```
```suggestion
			Determines which comparison operator is used when testing depth. See [enum DepthFunction].
			[b]Note:[/b] Changing [member depth_function] to a non-default value only has a visible effect when used on a transparent material, or a material that has [member depth_draw_mode] set to [constant DEPTH_DRAW_DISABLED].
```

apples commented:
Updated:

- Rebased to latest master.
- Incorporated Calinou's suggested documentation change.
- Renamed and reordered some enum values, to better match the underlying graphics API enums, and also to be more in line with the expected changes for https://github.com/godotengine/godot-proposals/issues/7174.
- Regarding the `glDepthMask` change from https://github.com/godotengine/godot/commit/ed0c378bda98d41288c0f9eeb31707590853e215: that change shouldn't affect this PR, so I left it all as-is.

apples commented:
Previous change broke Visual Shaders, pushed a small fix that involved moving the `LESS_OR_EQUAL` operator to slot 0 in all enums, since Visual Shaders assume 0 is the default value.

apples committed:
Add depth function for spatial materials

clayjohn commented:
I think this looks good from a design standpoint, I am also concerned about the impact of https://github.com/godotengine/godot/issues/73158. So I would like to discuss making the changes described here https://github.com/godotengine/godot/issues/73158#issuecomment-1430123182 in a future rendering meeting.

I'm worried about the usability of this PR. Switching between depth draw modes can be greatly impacted by the order in which nodes are draw (i.e if I render an object with depth_draw GREATER, but it is drawn first, it just won't get drawn. ``render_priority`` can help, but it can't sort between TAA and non-TAA passes or between opaque and transparent, so you can still get into trouble. 

We may want to hold off on this until we have the ability to run multiple opaque passes (and clear depth etc. between passes). Otherwise I fear this feature is going to feel a bit user hostile and difficult to control for anything except simple examples.

apples commented:
Closing as superseded by #80710.
